src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs

```diff
@@ -260,5 +261,239 @@ public static bool HasAllSet(this BitArray bitArray)
+                }
+
+                integral = span.Slice(0, i);
+
+                if (span[i] == '.')
+                {
+                    span = span.Slice(i + 1);
+#if NET
+                    i = span.IndexOfAny(s_exponent);
+#else
+                    i = span.IndexOfAny((byte)'e', (byte)'E');
+#endif
+
+                    if (i < 0)
+                    {
+                        fractional = span;
+                        exponent = 0;
+                        goto Normalize;
+                    }
+
+                    fractional = span.Slice(0, i);
+                }
+                else
+                {
+                    fractional = default;
+                }
+
+                Debug.Assert(span[i] is (byte)'e' or (byte)'E');
+                bool success = Utf8Parser.TryParse(span.Slice(i + 1), out exponent, out _);
+                Debug.Assert(success);
+
+            Normalize:
+                if (integral[0] == '0')
+                {
+                    // Normalize "0" to the empty span.
+                    Debug.Assert(integral.Length == 1, "Leading zeros not permitted in JSON numbers.");
+                    integral = default;
+                }
+
+                if (IndexOfFirstTrailingZero(fractional) is >= 0 and int iz)
+                {
+                    // Trim all trailing zeros from the fractional part.
+                    fractional = fractional.Slice(0, iz);
+                }
+
+                if (fractional.IsEmpty && IndexOfFirstTrailingZero(integral) is >= 0 and int fz)
+                {
+                    // There is no fractional part, trim all trailing zeros from
+                    // the integral part and increase the exponent accordingly.
+                    exponent += integral.Length - fz;
+                    integral = integral.Slice(0, fz);
+                }
+
+                // Normalize the exponent by subtracting the length of the fractional part.
+                exponent -= fractional.Length;
+
+                if (integral.IsEmpty)
+                {
+                    // Handle representations that only have a fractional component.
+
+                    if (IndexOfLastLeadingZero(fractional) is >= 0 and int lz)
+                    {
+                        // Trim all leading zeros from the fractional segment as
+                        // they have already been accounted for in the exponent.
+                        fractional = fractional.Slice(lz + 1);
+                    }
+
+                    if (fractional.IsEmpty)
+                    {
+                        // Normalize zero representations.
+                        isNegative = false;
+                        exponent = 0;
+                    }
+                }
+
+                static int IndexOfLastLeadingZero(ReadOnlySpan<byte> span)
+                {
+                    if (span.IsEmpty)
+                    {
+                        return -1;
+                    }
```

Comment:
Is empty expected to be common? If not, you could just delete this check; from the rest of the logic, it looks like -1 for empty naturally falls out.

Reply:
It should be common enough, empty spans are used to integers or values that only have a fractional part. I think this can be removed, at least for the first helper.